File: uniformity/R/uniformity.R
# Given a matrix of coordinate vectors (rows), calculate the Minimum Spanning
# Tree.
# Implemented for memory efficiency: does NOT store distance matrix.
# So memory use is O(n), not O(n^2). Time complexity O(n^2).
euclidMST <- function(mat) {
	.C("uniformity_mst", as.double(mat), as.integer(dim(mat)[1]), as.integer(dim(mat)[2]), result=integer(dim(mat)[1] - 1))$result
}

## Test that "euclidMST" is correctly implemented:
# library(vegan)
# all(euclidMST(bench) == spantree(dist(bench))$kid)

# Given a matrix of coordinate vectors (rows), calculate the nearest
# neighbours.
# Implemented for memory efficiency: does NOT store distance matrix.
# So memory use is O(n), not O(n^2). Time complexity O(n^2).
euclidNN <- function(mat) {
	res <- .C("uniformity_nn", as.double(mat), as.integer(dim(mat)[1]), as.integer(dim(mat)[2]), nearest=integer(dim(mat)[1]), sqdist=double(dim(mat)[1]))
	list(nearest=res$nearest, distance=sqrt(res$sqdist))
}

## Test that "euclidNN" is correctly implemented:
# shortest1 <- apply(as.matrix(dist(bench)) + diag(NA, dim(bench)[1]), 1, function(x) { min(x, na.rm=T) })
# shortest2 <- euclidNN(bench)$distance
# all(shortest1 == shortest2)

# Friedman-Rafsky test that two samples are from the same population (null).
# [Friedman and Rafsky, 1979 and Smith and Jain, 1984]
testUniformMST <- function(bench, test) {
	# Generate connectivity list: if st[i] = j, there is an edge (i + 1) -- (j)
	st <- euclidMST(rbind(bench, test))
	# Transform the list to obtain true edge-pairs
	st <- t(sapply(1:length(st), function(i) { c(i + 1, st[i]) }))

	N <- dim(bench)[1]
	M <- dim(test)[1]
	L <- M + N
	# Expected value of T (cross-connectivity)
	E <- 2*M*N/L
	# Calculate cross-connectivity between the sets
	T <- sum(apply(st, 1, function(e) { if ((e[1] <= N && e[2] > N) || (e[1] > N && e[2] <= N)) { 1 } else { 0 } }))

	# C is the number of edge pairs that share a vertex. Calculating it directly (like this) is slow:
	## C <- sum(sapply(1:(dim(st)[1] - 1), function(i) { e <- st[i,]; rem <- tail(st, -i); sum(rem[,1] == e[1] | rem[,1] == e[2] | rem[,2] == e[1] | rem[,2] == e[2]) }))
	# But it is related to the degree of the vertices, determined as:
	d <- table(c(st[,1], st[,2]))
	C <- sum(d * (d - 1) / 2)

	# Variance of T
	V <- (2 * M * N / (L * (L - 1))) * ((2 * M * N - L) / L + (C - L + 2) / ((L - 2) * (L - 3)) * (L * (L - 1) - 4 * M * N + 2))

	# values that can be then Z-tested(?)
	(T - E)/sqrt(V)
}

# Test uniformity using the Coefficient of Variation of the nearest-neighbour
# distances
testUniformCOV <- function(data) {
	# Calculate minimum inter-vertex distances
	gamma <- euclidNN(data)$distance

	# Calculate COV measure
	sd(gamma)/mean(gamma)
}

# Test uniformity using the Mesh Ratio
testUniformMR <- function(data) {
	# Calculate minimum inter-vertex distances
	gamma <- euclidNN(data)$distance

	# Calculate MR measure
	max(gamma) / min(gamma)
}

# Test uniformity by normalized error in mean (centroid)
testUniformComponentErrors <- function(data, centroid) {
	abs(apply(data, 2, mean) - centroid) / apply(data, 2, sd)
}

testUniformRatioError <- function(data, centroid) {
	n <- length(centroid)
	rr <- centroid[1] / centroid[n] # The real ratio
	est <- apply(data, 2, mean)
	er <- est[1] / est[n] # The estimated ratio
	abs(rr - er) / rr # Relative error
}

File: uniformity/src/uniformity.c
#include <R.h>
#include <math.h>
#include <stdlib.h>

typedef struct Matrix {
	double * const data;
	int const nRow;
	int const nCol;
} Matrix;

/**
 * @param i Row index.
 * @param j Column index.
 */
inline double *get(Matrix *m, int i, int j) {
	return m->data + j * (m->nRow) + i;
}

inline void swap(int *arr, int i, int j) {
	int tmp = arr[i];
	arr[i] = arr[j];
	arr[j] = tmp;
}

/**
 * Let the rows of the matrix represent points in Euclidians space.
 * Calculate the (squared) distance between the p0-th and p1-st row.
 * Avoids calculating the sqrt() since the squared distance preserves ordering.
 */
double uniformity_sqdist(Matrix *m, int p0, int p1) { 
	double dist = 0.0;
	for (int j = 0; j < m->nCol; ++j) {
		double const d = *get(m, p0, j) - *get(m, p1, j);
		dist += d * d;
	}
	return dist;
} 

/**
 * Calculate the minimum spanning tree of the given set of points.
 * Points are given as coordinate vectors in Euclidian space. The MST is
 * calculated using Prim's algorithm and returned as a set of edges.
 * @param points A matrix in which each row represents a coordinate vector.
 * @param nDim The dimension of the Euclidian space.
 * @param nPoint The number of points.
 * @param edges The return value: predecessor list (starting with 2nd point).
 */
void uniformity_mst(double *points, int *nPoint, int *nDim, int *edges) {
	Matrix m = { points, *nPoint, *nDim };

	if (m.nRow == 0) return;

	int *q = malloc(sizeof(int) * m.nRow); // Index list of visited / non-visited vertices
	int *n = malloc(sizeof(int) * m.nRow); // Nearest neighbour list
	double *d = malloc(sizeof(double) * m.nRow); // Shortest-distance list

	for (int i = 0; i < m.nRow; ++i) {
		q[i] = i;
		n[i] = 0;
		d[i] = uniformity_sqdist(&m, 0, i);
	}
	
	for (int i = 1; i < m.nRow; ++i) {
		// Find the currently closest point
		int closest = i;
		double minDist = d[q[i]];
		for (int p = i + 1; p < m.nRow; ++p) {
			if (d[q[p]] < minDist) {
				closest = p;
				minDist = d[q[p]];
			}
		}
		edges[q[closest] - 1] = n[q[closest]] + 1; // Add edge (R indices)
		swap(q, i, closest); // Move closest point to list of visited

		// Calculate distances from newly visited point
		for (int p = i + 1; p < m.nRow; ++p) {
			double dist = uniformity_sqdist(&m, q[i], q[p]);
			if (dist < d[q[p]]) {
				d[q[p]] = dist;
				n[q[p]] = q[i];
			}
		}
	}

	free(q);
	free(n);
	free(d);
}

/**
 * For each point in the given set of points, find its nearest neighbour and
 * record the distance to the nearest neighbour.
 * @param points A matrix in which each row represents a coordinate vector.
 * @param nDim The dimension of the Euclidian space.
 * @param nPoint The number of points.
 * @param nearest For each vertex, the nearest vertex.
 * @param dist For each vertex, the (squared) distance to the nearest vertex.
 */
void uniformity_nn(double *points, int *nPoint, int *nDim, int *nearest, double *dist) {
	Matrix m = { points, *nPoint, *nDim };

	for (int i = 0; i < m.nRow; ++i) {
		nearest[i] = NA_INTEGER;
		dist[i] = NA_REAL;
	}

	for (int i = 0; i < m.nRow; ++i) {
		for (int j = i + 1; j < m.nRow; ++j) {
			double d = uniformity_sqdist(&m, i, j);
			if (ISNA(dist[i]) || d < dist[i]) {
				dist[i] = d;
				nearest[i] = j;
			}
			if (ISNA(dist[j]) || d < dist[j]) {
				dist[j] = d;
				nearest[j] = i;
			}
		}
	}
}
